apollo-router/src/plugins/authentication/tests.rs

```diff
@@ -1357,6 +1367,180 @@ async fn issuer_check() {
     }
 }
 
+#[tokio::test]
+async fn audience_check() {
+    let signing_key = SigningKey::random(&mut OsRng);
+    let verifying_key = signing_key.verifying_key();
+    let point = verifying_key.to_encoded_point(false);
+
+    let encoding_key = EncodingKey::from_ec_der(&signing_key.to_pkcs8_der().unwrap().to_bytes());
+
+    let jwk = Jwk {
+        common: CommonParameters {
+            public_key_use: Some(PublicKeyUse::Signature),
+            key_operations: Some(vec![KeyOperations::Verify]),
+            key_algorithm: Some(KeyAlgorithm::ES256),
+            key_id: Some("hello".to_string()),
+            ..Default::default()
+        },
+        algorithm: AlgorithmParameters::EllipticCurve(EllipticCurveKeyParameters {
+            key_type: EllipticCurveKeyType::EC,
+            curve: EllipticCurve::P256,
+            x: BASE64_URL_SAFE_NO_PAD.encode(point.x().unwrap()),
+            y: BASE64_URL_SAFE_NO_PAD.encode(point.y().unwrap()),
+        }),
+    };
+
+    let manager = make_manager(
+        &jwk,
+        None,
+        Some(HashSet::from(["hello".to_string(), "goodbye".to_string()])),
+    );
+
+    // No audience
+    let token = encode(
+        &jsonwebtoken::Header::new(Algorithm::ES256),
+        &Claims {
+            sub: "test".to_string(),
+            exp: get_current_timestamp(),
+            aud: None,
+            iss: None,
+        },
+        &encoding_key,
+    )
+    .unwrap();
+
+    let request = supergraph::Request::canned_builder()
+        .header(http::header::AUTHORIZATION, format!("Bearer {token}"))
+        .build()
+        .unwrap();
+
+    let mut config = JWTConf::default();
+    config.sources.push(Source::Header {
+        name: super::default_header_name(),
+        value_prefix: super::default_header_value_prefix(),
+    });
+    match authenticate(&config, &manager, request.try_into().unwrap()) {
+        ControlFlow::Break(res) => {
+            assert_eq!(res.response.status(), StatusCode::UNAUTHORIZED);
+            let body = res.response.into_body().collect().await.unwrap();
+            let body = String::from_utf8(body.to_bytes().to_vec()).unwrap();
+            assert_eq!(
+                body,
+                "{\"errors\":[{\"message\":\"Invalid audience: the token's `aud` was 'null', but 'goodbye, hello' was expected\",\"extensions\":{\"code\":\"AUTH_ERROR\"}}]}"
```

Comment:
would be nicer to compare the parsed json value here (so there's no reliance on key order for example), but not a huge deal.